src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -810,31 +811,62 @@ private LanguageExpression ConvertBinary(BinaryOperationSyntax syntax)
 
         private LanguageExpression ConvertUnary(UnaryOperationSyntax syntax)
         {
-            LanguageExpression convertedOperand = ConvertExpression(syntax.Expression);
-
             switch (syntax.Operator)
             {
                 case UnaryOperator.Not:
+                    LanguageExpression convertedOperand = ConvertExpression(syntax.Expression);
                     return CreateFunction("not", convertedOperand);
 
                 case UnaryOperator.Minus:
-                    if (convertedOperand is JTokenExpression literal && literal.Value.Type == JTokenType.Integer)
+                    if (syntax.Expression is IntegerLiteralSyntax integerLiteral)
                     {
-                        // invert the integer literal
-                        int literalValue = literal.Value.Value<int>();
-                        return new JTokenExpression(-literalValue);
+                        //do codegen logic
+                        LanguageExpression integerExpression = ConvertInteger(integerLiteral, true);
+                        return integerExpression;
                     }
 
                     return CreateFunction(
                         "sub",
                         new JTokenExpression(0),
-                        convertedOperand);
+                        ConvertExpression(syntax.Expression));
 
                 default:
                     throw new NotImplementedException($"Cannot emit unexpected unary operator '{syntax.Operator}.");
             }
         }
 
+        // for 32 bit integers, we return the integer value
+        // for values outside that range, return the FunctionExpression
+        private LanguageExpression ConvertInteger(IntegerLiteralSyntax integerSyntax, bool minus)
+        {
+            long integerValue = (long)integerSyntax.Value;
+
+            if (minus)
+            {
+                // integerSyntax.Value is always positive, so for the most negative signed 32 bit integer -2,147,483,648
+                // we would compare its positive token (2,147,483,648) to int.MaxValue (2,147,483,647) + 1
+                if (integerSyntax.Value > (ulong)int.MaxValue + 1)
+                {
+                    long negativeIntegerValue = -1 * integerValue;
+                    return CreateFunction("json", new JTokenExpression(negativeIntegerValue.ToInvariantString()));
+                }
+                else
+                {
+                    return new JTokenExpression((int)integerSyntax.Value);
```

Comment:
Don't forget to negate the value here! The order that you do things is important to avoid overflows, so you'll need to ensure you negate before casting.